### ISSUE TITLE
Crash during AST cleanup

### ISSUE THREAD

**Issue by Tropicao** [python/cpython#117644]
# Crash report

### What happened?

Hello,
I am encountering issues with python cross-built for an ARM platform, and got a systematic crash with a pretty simple reproducer:
```bash
echo 'print("foo")'|python
```
(it happens too if the `print("foo")` is stored in a python script and passed to the interpreter)

I have to mention that the used cpython binary is not a "pure" upstream version, but the one packaged by [Buildroot ](https://buildroot.org/) on tag 2024.02.1 (so the used sources come from [package 3.11.8 from python public ftp server](https://www.python.org/ftp/python/3.11.8/Python-3.11.8.tar.xz), and are slightly patched by buildroot). I understand that I may then be redirected to buildroot mailing lists to reproduce the issue, but since I am not able to provide a simple buildroot defconfig reproducer (the issue does not happen on an "isolated", minimal build but on a whole firmware build), I hope to get some guidance here about my observations.

When analyzing the core file/trying to debug live the binary, the crash systematically occurs in `_PyAST_Fini` function:
```
#0  0xb6c691e8 in Py_DECREF (op=<optimized out>) at Python/Python-ast.c:192
#1  _PyAST_Fini (interp=interp@entry=0xb6ee9ce0 <_PyRuntime+31096>) at Python/Python-ast.c:192
#2  0xb6ce5458 in interpreter_clear (interp=0xb6ee9ce0 <_PyRuntime+31096>, tstate=0xb6ef6f48 <_PyRuntime+84960>) at Python/pystate.c:431
#3  0xb6ce58c4 in _PyInterpreterState_Clear (tstate=<optimized out>) at Python/pystate.c:472
#4  0xb6cdc1fc in finalize_interp_clear (tstate=tstate@entry=0xb6ef6f48 <_PyRuntime+84960>) at Python/pylifecycle.c:1708
#5  0xb6cdfc54 in Py_FinalizeEx () at Python/pylifecycle.c:1899
#6  Py_FinalizeEx () at Python/pylifecycle.c:1751
#7  0xb6d0d734 in Py_RunMain () at Modules/main.c:682
#8  0xb6d0e188 in pymain_main (args=args@entry=0xbe91fd24) at Modules/main.c:710
#9  0xb6d0e2b4 in Py_BytesMain (argc=<optimized out>, argv=<optimized out>) at Modules/main.c:734
#10 0xb69518a8 in __libc_start_call_main () from ./lib/libc.so.6
#11 0xb69519b0 in __libc_start_main_impl () from ./lib/libc.so.6
#12 0x004d9738 in _start ()
```
When analyzing a bit further, I observe that `state` in `_PyAST_Fini` has indeed a lot of uninitialized/bogus objects pointers starting from an arbitrary offset:
```
[...]
  arg = 0x0,
  arg_type = 0x0,
  args = 0x0,
  argtypes = 0x0,
  arguments_type = 0x0,
  asname = 0x0,
  ast = 0x0,
  attr = 0x0,
  bases = 0x0,
  body = 0x0,
  boolop_type = 0x0,
  cases = 0x0,
  cause = 0x0,
  cls = 0x0,
  cmpop_type = 0x0,
  col_offset = 0x0,
  comparators = 0xe1a0300b, # The crash indeed occurs when executing Py_CLEAR(state->comparators);
  comprehension_type = 0xe1a01006,
  context_expr = 0xe1a00005,
  conversion = 0xe58d7000,
  ctx = 0xeb009dff,
  decorator_list = 0xe3500000,
  defaults = 0x1a000054,
  elt = 0xe3e06000,
  elts = 0xe3a08000,
  end_col_offset = 0xea000025,
  end_lineno = 0xe5952068,
  exc = 0xe3520002,
  excepthandler_type = 0xe595201c,
  expr_context_type = 0xa000004,
  expr_type = 0xe3520000,
  finalbody = 0xa00000d
[...]
```
I tried to follow the corresponding initialization path and kind of understood the following:
- the `ast` structure here is embedded in a `PyInterpreterState`, itself coming from a `_PyRuntimeState`
- the corresponding `_PyRuntimeState` struct used here is in fact the global variable `_PyRuntime`, and is initialized by copying into it the content of `_PyRuntimeState_INIT`
- `_PyRuntimeState_INIT` has a few fields initialized. The `ast` structure, embedded in the `_main_interpreter` field (initialized with `_PyInterpreterState_INIT` content) does not receive a default value

Now, not knowing anything about cpython architecture, I am wondering whether any of the following thing should be done:
- initialize the AST structure in `_PyInterpreterState_INIT` ?
- since this structure has an `initialized` field, check this field and skip `_PyAST_Fini` if it is not 1 ?
- none of above, and I am missing something bigger ? Or already fixed in other versions ?

### CPython versions tested on:

3.11

### Operating systems tested on:

Linux

### Output from running 'python -VV' on the command line:

Python 3.11.8 (main, Mar 29 2024, 14:05:00) [GCC 12.3.0]

**Comment by vstinner** [comment]
Can you reproduce the issue if you build Python from source using `./configure && make`? I don't know if you can build directly on ARM. It seems like other users are not affected.

Python 3.11 is no longer supported: https://devguide.python.org/versions/ Can you reproduce the issue with Python 3.12, or even 3.13?

**Comment by Tropicao** [comment]
I eventually observed that the issue was not occurring when upgrading to v3.11.9. Since I had a limited time budget to investigate this, and since the issue does not seem to reproduce on 3.11.9, I moved on and used this version. I did not have the opportunity to test on higher versions since then.
If no one else is encountering this specific crash, I suggest to close the issue. If I encounter the issue on newer version, I'll open a new issue.
